### PR TITLE
Add recordingrule paramater

### PR DESCRIPTION
**Pull request by haoqing0110**
Ref https://github.com/open-cluster-management/backlog/issues/11950

### CODE REVIEW

File: cmd/metrics-collector/main.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ func main() {
 
 	// TODO: more complex input definition, such as a JSON struct
 	cmd.Flags().StringArrayVar(&opt.Rules, "match", opt.Rules, "Match rules to federate.")
+	cmd.Flags().StringArrayVar(&opt.RecordingRules, "recordingrule", opt.RecordingRules, "Define recording rules to reduce metrics before sending them, one rule per line.")
```
Comment (marcolan018) [inline cmd/metrics-collector/main.go]: recording rule is to generate new metrics based on specified query expression. 

" reduce metrics before sending them" is the one purpose we can use recording rule to achieve

File: pkg/forwarder/forwarder.go [hunk 2]
```diff
@@ -368,5 +385,72 @@ func (w *Worker) forward(ctx context.Context) error {
 			rlogger.Log(w.logger, rlogger.Warn, "msg", failedStatusReportMsg, "err", err)
 		}
 	}
+
+	elapsed := time.Since(fStart)
+	rlogger.Log(w.logger, rlogger.Info, "forward_total_time", elapsed)
```
Comment (marcolan018) [inline pkg/forwarder/forwarder.go]: don't need the api invocation time, that's only for debug/investigation before

File: pkg/forwarder/forwarder.go [hunk 3]
```diff
@@ -298,26 +315,26 @@ func (w *Worker) Run(ctx context.Context) {
...
+		return err
 	}
-	from.RawQuery = v.Encode()
 
-	req := &http.Request{Method: "GET", URL: from}
-	families, err := w.fromClient.Retrieve(ctx, req)
+	rfamilies, err := w.getRecordingMetrics(ctx)
 	if err != nil {
-		statusErr := w.status.UpdateStatus("Degraded", "Degraded", "Failed to retrieve metrics")
+		statusErr := w.status.UpdateStatus("Degraded", "Degraded", "Failed to retrieve recording metrics")
```
Comment (marcolan018) [inline pkg/forwarder/forwarder.go]: if failed to get result for recording rule, should we return an error, or return the federate result only?
I guess latter?

File: pkg/forwarder/forwarder.go [hunk 4]
```diff
@@ -368,5 +385,72 @@ func (w *Worker) forward(ctx context.Context) error {
...
+		// reset query from last invocation, otherwise match rules will be appended
+		from.RawQuery = ""
+		v := w.from.Query()
+		v.Add("query", rquery)
+		from.RawQuery = v.Encode()
+
+		req := &http.Request{Method: "GET", URL: from}
+		rfamilies, err := w.fromClient.RetrievRecordingMetrics(ctx, req, rname)
+		if err != nil {
+			return families, err
```
Comment (marcolan018) [inline pkg/forwarder/forwarder.go]: if err, I think use "continue" is better

File: pkg/forwarder/forwarder.go [hunk 5]
```diff
@@ -368,5 +385,72 @@ func (w *Worker) forward(ctx context.Context) error {
...
+		} else {
+			families = append(families, rfamilies...)
+		}
+	}
+
+	elapsed := time.Since(start)
+	rlogger.Log(w.logger, rlogger.Info, "recording_query_time", elapsed)
```
Comment (marcolan018) [inline pkg/forwarder/forwarder.go]: also don't need the invocation time

File: pkg/forwarder/forwarder.go (hunk 3 above)
Comment (haoqing0110) [reply pkg/forwarder/forwarder.go]: fixed, won't return error when getRecordingMetrics has error.

File: cmd/metrics-collector/main.go (hunk 1 above)
Comment (haoqing0110) [reply cmd/metrics-collector/main.go]: fixed

File: pkg/forwarder/forwarder.go (hunk 2 above)
Comment (haoqing0110) [reply pkg/forwarder/forwarder.go]: fixed

File: pkg/forwarder/forwarder.go (hunk 4 above)
Comment (haoqing0110) [reply pkg/forwarder/forwarder.go]: fixed, will only print warning and continue

File: pkg/forwarder/forwarder.go (hunk 5 above)
Comment (haoqing0110) [reply pkg/forwarder/forwarder.go]: fixed

**Comment by haoqing0110** [comment]
@marcolan018  thanks, there's some conflict with PR https://github.com/open-cluster-management/metrics-collector/pull/20, may also need you help confirm the changes in latest commit https://github.com/open-cluster-management/metrics-collector/pull/24/commits/3a48f0ae818b79b7e1dab95ddf8b8904d107f9b6

**Comment by clyang82** [comment]
Agreed with @marcolan018 if simulate is true, just return simulate data. if we need to handle recording rule, it should be inside of simulator.

**Comment by haoqing0110** [comment]
close as too much conflicts and commits in this PR. Plz refer to new PR https://github.com/open-cluster-management/metrics-collector/pull/27

**Pull request closed without merging**